Add vector construction utilities

mcol opened the pull request:
## Summary

This adds the following function signatures (well the underlying code to implement them):
- `vector constant_vector(int K, real c)`
- `vector one_hot_vector(int K, int k)`
- `vector set_spaced_vector(int K, real low, real high)`
- `vector uniform_simplex(int K)`
- `vector zeros_vector(int K)`
- `matrix identity_matrix(int K)`

This departs slightly from the signatures mentioned in #692 for `constant_vector`, `one_hot_vector` and `set_spaced_vector`, in that `K` is always the first parameter, which is also what Eigen uses.

## Tests

Added.

## Side Effects

None.

## Checklist

- [X] Math issue #692

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

bob-carpenter commented:
I'll review this one.

bob-carpenter commented on test/unit/math/prim/fun/one_hot_vector_test.cpp:
```diff
@@ -0,0 +1,38 @@
+
+TEST(MathFunctions, one_hot_vector_throw) {
+  using stan::math::one_hot_vector;
+  int K = 5;
```
This needs a K = 0, k = 1 test to make sure that K = 0 is being caught as an exception.  Or the tested K needs to be put in a loop fro 0 to 5 or something.

bob-carpenter commented on test/unit/math/prim/fun/ones_vector_test.cpp:
```diff
@@ -0,0 +1,31 @@
+  using Eigen::VectorXd;
+  using stan::math::ones_vector;
+
+  VectorXd u0 = ones_vector(0);
```
[optional]
This can also all go in a loop:

```
for (int K = 0; K < 5; ++K) {
  Eigen::VectorXd y = Eigen::Constant(1, K);  // or whatever the right syntax for this is
  expect_float_eq(y, ones_vector(K));
}
```

It doesn't matter if that's how you implemented the `ones_vector` function as it's clearly the right thing to test.  The point is to have the test be clear and right.

bob-carpenter commented on test/unit/math/prim/fun/ones_vector_test.cpp:
```diff
@@ -0,0 +1,31 @@
+
+TEST(MathFunctions, ones_vector_throw) {
+  using stan::math::ones_vector;
+
```
[optional]
I'd rather see less vertical space when it'd still be readable without.
Also, it'd save another line to just put the `stan::math` into the test call.

bob-carpenter commented on test/unit/math/prim/fun/set_spaced_vector_test.cpp:
```diff
@@ -0,0 +1,52 @@
+TEST(MathFunctions, set_spaced_vector) {
+  using Eigen::VectorXd;
+  using stan::math::set_spaced_vector;
+
```
Something needs to test `K = 0` and `K = 1` cases.

bob-carpenter commented on test/unit/math/prim/fun/uniform_simplex_test.cpp:
```diff
@@ -0,0 +1,31 @@
+  using Eigen::VectorXd;
+  using stan::math::uniform_simplex;
+
+  VectorXd u0 = uniform_simplex(0);
```
This should throw, not return a size zero result, because a size zero vector is not a simplex (entries don't sum to 1).

bob-carpenter commented on test/unit/math/prim/fun/uniform_simplex_test.cpp:
```diff
@@ -0,0 +1,31 @@
+
+TEST(MathFunctions, uniform_simplex_throw) {
+  using stan::math::uniform_simplex;
+
```
This should include 0 case.

bob-carpenter commented on test/unit/math/prim/fun/zeros_vector_test.cpp:
```diff
@@ -0,0 +1,31 @@
+TEST(MathFunctions, zeros_vector) {
+  using Eigen::VectorXd;
+  using stan::math::zeros_vector;
+
```
[optional]
This is another one to do in a loop.  Really, so are most of the others.

mcol commented on test/unit/math/prim/fun/identity_matrix_test.cpp:
```diff
@@ -0,0 +1,36 @@
+  int K = 3;
+  MatrixXd u2 = identity_matrix(K);
+  MatrixXd m2(K, K);
+  m2 << 1, 0, 0, 0, 1, 0, 0, 0, 1;
```
I think I had it the way you are suggesting, then clang-format said no. :)

bob-carpenter replied:
OK.  We can't fight clang format on this, though that's too bad for readability.

mcol commented on stan/math/prim/fun/ones_vector.hpp:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @param K length of the vector
+ * @return A vector of length K with all elements initialised to 1.
+ */
```
We are very inconsistent in the formatting of doxygen comments. Most I've seen don't use a period, but I tend to format the `@return` as a full sentence, so I capitalize and put a period at the end.

bob-carpenter replied:
You're right that the doc on the parameters and return values is inconsistent.  My comment was about the top-level doc string, not the return or param doc!

But while we're on that topic... The typographically correct thing to do for `@param` and `@return` is to use a noun phrase that describes the input or output and thus not use any capitalization or periods.

In the example above, "A vector of length K with all elements initialised to 1" is a noun phrase, not a sentence.  You can tell because there's no main verb.  The only verb is `initialised`, which appears in the subordinate clause "with all elements initialized to 1", which is a modifier of the head noun "vector".

mcol commented on stan/math/prim/fun/set_spaced_vector.hpp:
```diff
@@ -0,0 +1,32 @@
+namespace math {
+
+/**
+ * Return a vector of linearly spaced elements
```
For K=0, the vector will be empty, for K=1 it returns high, which  is what Eigen does. I'll add this to docs and tests.

mcol commented on test/unit/math/prim/fun/constant_vector_test.cpp:
```diff
@@ -0,0 +1,33 @@
+  VectorXd u1 = constant_vector(1, c);
+  VectorXd v1(1);
+  v1 << c;
+  EXPECT_EQ(1, u1.size());
```
The current version works already on vectors and matrices. What's the advantage of templating it?

bob-carpenter replied:
Which current version?  What I want is a test function to compare Eigen matrices, row vectors, and vectors so we don't have to keep testing sizes and looping.

mcol commented:
Thanks Bob! I've addressed almost everything, I've answered you for things I was not sure.

As suggested I replaced most tests with loops, even if they used the same functions used in the headers being tested, but I left the tests for `set_spaced_vector` unchanged as they convey more clearly what linearly spaced means.

I've also changed the check in `one_hot_vector` not to allow 0-size vectors: this is because the function would throw anyway because on a size 0 vector there's no way we can specify an index of at least 1. In either case we throw, but now it's at least more obvious.

bob-carpenter commented:
> For K=0, the vector will be empty, for K=1 it returns high, which is what Eigen does. I'll add this to docs and tests.

Thanks!  I had no idea what it did.
